### PR TITLE
Navigation Signals in PPR

### PR DESCRIPTION
**Pull request by wyattjoh**
### What

This adds support for navigation signals like `notFound()` and `redirect(url)` when Partial Prerendering has been enabled.

### Why

Navigation API's like `notFound()` and `redirect(url)` throw errors in order to interrupt the rendering of components. When a page both invokes API's that cause the render to be marked as dynamic (like `unstable_noStore()`) and also a navigation API, these errors may race to the end. In the case where the navigation error does not beat out the error emitted by dynamic API's will still trigger the detection warning that's present to warn you about situations where you may have accidentally caught the error.

### How

This resolves this issue by explicitly checking for navigation signals (errors) thrown during the render, and not displaying the "caught dynamic API" error and console warning.

Closes NEXT-2037

### CODE REVIEW

**Comment by wyattjoh** [comment]
<b>In this stack:</b> <code>🧭 Enhancing Navigation Signals in PPR Stack</code>
* `canary`
  * **#60450** <a href="https://app.graphite.dev/github/pr/vercel/next.js/60450?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @wyattjoh and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| buildDuration | 12.8s | 12.8s | N/A |
| buildDurationCached | 7.2s | 6.2s | N/A |
| nodeModulesSize | 199 MB | 200 MB | ⚠️ +22.8 kB |
| nextStartRea..uration (ms) | 434ms | 435ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | N/A |
| 433-HASH.js gzip | 28.5 kB | 28.6 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 243 B | N/A |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 45.2 kB | 45.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| index.html gzip | 529 B | 526 B | N/A |
| link.html gzip | 541 B | 539 B | N/A |
| withRouter.html gzip | 525 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.8 kB | 93.8 kB | N/A |
| page.js gzip | 148 kB | 148 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 625 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.5 kB | 37.5 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/navigation-signals-ppr | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 169 kB | 169 kB | N/A |
| app-page-exp..prod.js gzip | 95.3 kB | 95.3 kB | N/A |
| app-page-tur..prod.js gzip | 96 kB | 96 kB | N/A |
| app-page-tur..prod.js gzip | 90.6 kB | 90.6 kB | N/A |
| app-page.run...dev.js gzip | 142 kB | 142 kB | N/A |
| app-page.run..prod.js gzip | 89.9 kB | 89.9 kB | N/A |
| app-route-ex...dev.js gzip | 24.1 kB | 24.1 kB | N/A |
| app-route-ex..prod.js gzip | 16.7 kB | 16.8 kB | N/A |
| app-route-tu..prod.js gzip | 16.7 kB | 16.8 kB | N/A |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | N/A |
| app-route.ru...dev.js gzip | 23.5 kB | 23.5 kB | N/A |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | N/A |
| pages-api-tu..prod.js gzip | 9.38 kB | 9.38 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.65 kB | 9.65 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.5 kB | 49.5 kB | ✓ |
| Overall change | 144 kB | 144 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>433-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: fda552c26d924183c84a0f501dc19627d560a3be

**Comment by ijjk** [comment]
## Failing test suites 
Commit: fda552c26d924183c84a0f501dc19627d560a3be

`pnpm test-dev test/e2e/app-dir/ppr-full/ppr-full.test.ts` 
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found > should have correct headers
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found/dynamic > should have correct headers
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found/dynamic > should cache the static part
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect > should have correct headers
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect/dynamic > should have correct headers
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect/dynamic > should cache the static part

<details>
<summary>Expand output</summary>

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found/dynamic › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found/dynamic › should cache the static part

    expect(received).toBeLessThan(expected)

    Expected: < 500
    Received:   519

      309 |               // This is because the signal should throw and interrupt the
      310 |               // delay timer.
    > 311 |               expect(result.streamEnd - start).toBeLessThan(delay)
          |                                                ^
      312 |             })
      313 |           }
      314 |         })

      at Object.toBeLessThan (e2e/app-dir/ppr-full/ppr-full.test.ts:311:48)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect/dynamic › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect/dynamic › should cache the static part

    expect(received).toBeLessThan(expected)

    Expected: < 500
    Received:   522

      309 |               // This is because the signal should throw and interrupt the
      310 |               // delay timer.
    > 311 |               expect(result.streamEnd - start).toBeLessThan(delay)
          |                                                ^
      312 |             })
      313 |           }
      314 |         })

      at Object.toBeLessThan (e2e/app-dir/ppr-full/ppr-full.test.ts:311:48)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: fda552c26d924183c84a0f501dc19627d560a3be

`pnpm test-dev test/e2e/app-dir/ppr-full/ppr-full.test.ts` 
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found > should have correct headers
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found/dynamic > should have correct headers
- ppr-full > Navigation Signals > notFound() > for /navigation/not-found/dynamic > should cache the static part
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect > should have correct headers
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect/dynamic > should have correct headers
- ppr-full > Navigation Signals > redirect() > for /navigation/redirect/dynamic > should cache the static part

<details>
<summary>Expand output</summary>

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found/dynamic › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › notFound() › for /navigation/not-found/dynamic › should cache the static part

    expect(received).toBeLessThan(expected)

    Expected: < 500
    Received:   518

      309 |               // This is because the signal should throw and interrupt the
      310 |               // delay timer.
    > 311 |               expect(result.streamEnd - start).toBeLessThan(delay)
          |                                                ^
      312 |             })
      313 |           }
      314 |         })

      at Object.toBeLessThan (e2e/app-dir/ppr-full/ppr-full.test.ts:311:48)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect/dynamic › should have correct headers

    expect(received).toEqual(expected) // deep equality

    Expected: "s-maxage=31536000, stale-while-revalidate"
    Received: "no-store, must-revalidate"

      279 |               'text/html; charset=utf-8'
      280 |             )
    > 281 |             expect(res.headers.get('cache-control')).toEqual(
          |                                                      ^
      282 |               's-maxage=31536000, stale-while-revalidate'
      283 |             )
      284 |

      at Object.toEqual (e2e/app-dir/ppr-full/ppr-full.test.ts:281:54)

  ● ppr-full › Navigation Signals › redirect() › for /navigation/redirect/dynamic › should cache the static part

    expect(received).toBeLessThan(expected)

    Expected: < 500
    Received:   520

      309 |               // This is because the signal should throw and interrupt the
      310 |               // delay timer.
    > 311 |               expect(result.streamEnd - start).toBeLessThan(delay)
          |                                                ^
      312 |             })
      313 |           }
      314 |         })

      at Object.toBeLessThan (e2e/app-dir/ppr-full/ppr-full.test.ts:311:48)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

File: packages/next/src/server/app-render/app-render.tsx [hunk 1]
```diff
@@ -720,16 +721,22 @@ async function renderToHTMLOrFlightImpl(
     isRSCRequest &&
     req.headers[NEXT_ROUTER_PREFETCH_HEADER.toLowerCase()] !== undefined
 
+  // TODO: this should be moved to a higher point in the stack
```
Comment (ztanner) [inline packages/next/src/server/app-render/app-render.tsx:724]: TODO nit: I think ideally we can just remove the flight router state dependency for interception routes. It's only needed right now for the `makeGetDynamicParams` function, but seems like with some refactoring we could get it to behave.

**Review by ztanner — APPROVED** [review]

**Pull request merged**